Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.h

```diff
@@ -236,5 +308,181 @@ namespace Terrain
+            processJob->Start();
+        }
+    }
+
+    template<typename PerSurfacePointFunctionType>
+    inline void TerrainSystem::ProcessFromRegionAsync(PerSurfacePointFunctionType perSurfacePointFunction,
+        const AZ::Aabb& inRegion,
+        const AZ::Vector2& stepSize,
+        AzFramework::Terrain::SurfacePointRegionFillCallback perPositionCallback,
+        Sampler sampleFilter,
+        AZStd::shared_ptr<ProcessAsyncParams> params) const
+    {
+        if (!perPositionCallback)
+        {
+            return;
+        }
+
+        // ToDo: Determine the number of jobs to split the work into based on:
+        // 1. The number of available worker threads.
+        // 2. The desired number of jobs as passed in.
+        // 3. The size of the area being processed.
+        //
+        // Note: We are currently restricting the number of worker threads to one
+        // because splitting the work over multiple threads causes contention when
+        // locking various mutexes, resulting in slower overall wall time for async
+        // requests split over multiple threads vs one where all the work is done on
+        // a single thread. The latter is still preferable over a regular synchronous
+        // call because it is just as quick and prevents the main thread from blocking.
+        // Once the mutex contention issues have been addressed, we should come up with
+        // an algorithm to break up 'inRegion' into sub-regions (or lists of positions?)
+        // so that async calls automatically split the work between available job manager
+        // worker threads, unless the ProcessAsyncParams specifiy a desired number of jobs.
+        const int32_t numWorkerThreads = 1; // m_terrainJobManager->GetNumWorkerThreads();
+        const int32_t numJobsDesired = params ? params->m_desiredNumberOfJobs : -1;
+        const int32_t numJobs = (numJobsDesired > 0) ? AZStd::min(numWorkerThreads, numJobsDesired) : numWorkerThreads;
+        if (numJobs <= 0)
+        {
+            AZ_Warning("TerrainSystem", false, "No positions to process.");
+            return;
+        }
+
+        // Determine the desired terrain job context so the jobs can be cancelled.
+        TerrainJobContext* jobContext = (params && params->m_terrainJobContext) ?
+                                        params->m_terrainJobContext.get() :
+                                        m_defaultTerrainJobContext.get();
+
+        // Split the work across multiple jobs.
+        AZ_Assert(numJobs == 1, "We don't yet support breaking up regions."); // Temp until we figure out how to break up the region.
+        AZStd::shared_ptr<AZStd::atomic_int> numJobCompletionsRemaining = AZStd::make_shared<AZStd::atomic_int>(numJobs);
+        for (int32_t i = 0; i < numJobs; ++i)
+        {
+            // Define the job function using the sub region of positions to process.
+            const AZ::Aabb& subRegion = inRegion; // ToDo: Figure out how to break up the region.
```

Comment:
Suggestion: Maybe partition the subregions based on strips of Y values?
Ex: (0,0) -> (16,16), with 4 jobs does (0,0)->(16,3), (0,4)->(16,7), (0,8)->(16,11), (0,12)->(16,15).

Reply:
Going to leave this for later.

Reply:
I'd recommend long term trying to keep the regions squarish so the intermediate result may be more useful on its own. Could possibly even make this something passed in - for instance "here's a 1024x1024 region, please process in 64x64 chunks and let me know when each chunk is done".

Reply:
Unresolving this temporarily... I think Mike resolved this conversation right before I commented on it and I don't want my comment to get ignored 😄

Reply:
Good idea, although with the current design if the calling code wants a callback for each sub-region they would have to break up the work themselves and call this for each sub region with m_desiredNumJobs = 1

Your suggestion is certainly possible, we'd just have to think about how to structure the callbacks for each sub region.